### PR TITLE
USMON-824: USM Kafka support for produce record batching

### PR DESCRIPTION
**Pull request by DanielLavie**
### What does this PR do?
This PR enhances the USM Kafka parsing logic to recognize the count of Kafka records in a produce request.
The relevant section in the Kafka protocol specification can be found [here](https://kafka.apache.org/documentation/#recordbatch).

### Motivation
During testing in our staging environment, we observed that our Kafka producers utilize the Kafka batching mechanism, sending more than one Kafka produce record within a single request. This resulted in an accuracy discrepancy between us and DSM. As a resolution, we aim to incorporate support for this behavior.

### Additional Notes

This modification provides an opportunity to incorporate additional Kernel metrics, such as:

* Distribution of the number of records in a produce request
* Count of produce requests targeting more than a single partition

### Possible Drawbacks / Trade-offs
* In order to handle multiple records for a fetch request, it's necessary to parse Kafka produce responses. Currently, we lack the capability to parse Kafka responses, so this issue will need to be resolved in a separate PR.
* Reduced `KAFKA_BATCH_SIZE` from 30 to 28 due to an increase in the size of the `kafka_transaction_t` struct. This adjustment may potentially impact the occurrences of Kafka failed flushes in a negative manner.
* To begin with, for simplicity, we are focusing on scenarios where we produce to a single partition. This decision is based on our observations in the Datadog environment, where we haven't encountered instances of producers producing to multiple partitions within the same request.

### CODE REVIEW

File: pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h [hunk 1]
```diff
@@ -121,6 +121,48 @@ static __always_inline bool kafka_process(kafka_transaction_t *kafka_transaction
 
     log_debug("kafka: topic name is %s", kafka_transaction->topic_name);
 
+    switch (kafka_header.api_key) {
+    case KAFKA_PRODUCE:
+    {
+        READ_BIG_ENDIAN_WRAPPER(s32, number_of_partitions, skb, offset);
+        if (number_of_partitions > 1) {
+            log_debug("We have more than a single partition for produce request, we currently support only the first partition");
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h:129]: Either rephrase the log or the code, as the log says we support only the first partition, but the if clause above says we support messages with **a single** partition

so either we parse only the first partition (no matter how many partition exist) or the log should say we support messages with a single partition

File: pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h [hunk 2]
```diff
@@ -121,6 +121,48 @@ static __always_inline bool kafka_process(kafka_transaction_t *kafka_transaction
...
+            return false;
+        }
+        offset += sizeof(u32); // Skipping crc
+        offset += sizeof(s16); // Skipping attributes
+        offset += sizeof(s32); // Skipping last offset delta
+        offset += sizeof(s64); // Skipping base timestamp
+        offset += sizeof(s64); // Skipping max timestamp
+        offset += sizeof(s64); // Skipping producer id
+        offset += sizeof(s16); // Skipping producer epoch
+        offset += sizeof(s32); // Skipping base sequence
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h:149]: Didn't verify this, @vitkyrka can you review the PR as well and verify it against the wire-protocol documentation?

File: pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h (hunk 1 above)
Comment (DanielLavie) [reply pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h:129]: Good call, fixed

**Review by vitkyrka — APPROVED** [review]

File: pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h [hunk 3]
```diff
@@ -121,6 +121,48 @@ static __always_inline bool kafka_process(kafka_transaction_t *kafka_transaction
 
     log_debug("kafka: topic name is %s", kafka_transaction->topic_name);
 
+    switch (kafka_header.api_key) {
+    case KAFKA_PRODUCE:
+    {
+        READ_BIG_ENDIAN_WRAPPER(s32, number_of_partitions, skb, offset);
+        if (number_of_partitions > 1) {
```
[the remark this answers is not in the document]
Comment (DanielLavie) [reply pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h:128]: This should be part of the classification so I think it's redundant here, we should add this to the classification logic

File: pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h [hunk 4]
```diff
@@ -121,6 +121,48 @@ static __always_inline bool kafka_process(kafka_transaction_t *kafka_transaction
...
+            return false;
+        }
+        offset += sizeof(s32); // Skipping Partition ID
+        offset += sizeof(s32); // Skipping record batch (message set in wireshark) size in bytes
+        offset += sizeof(s64); // Skipping record batch baseOffset
+        offset += sizeof(s32); // Skipping record batch batchLength
+        offset += sizeof(s32); // Skipping record batch partitionLeaderEpoch
+        READ_BIG_ENDIAN_WRAPPER(s8, magic_byte, skb, offset);
+        if (magic_byte != 2) {
+            log_debug("Got magic byte != 2, the protocol state it should be 2");
```
[the remark this answers is not in the document]
Comment (DanielLavie) [reply pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h:139]: The format of the Message set is different in older versions, so printing the value alone doesn't reflect the magic in this case. However, I'll add a note about the older versions

**Pull request merged**